File: space-game/Core/Classes/Mass.vb
Namespace Classes
    Public Class Mass
        Implements IEquatable(Of Mass)

        Private ReadOnly _kilograms As Double

        Const KilogramsInSolarMass As Double = 1.98855 * 10 ^ 30
        Const KilogramsInEarthMass As Double = 5973.6 * 10 ^ 21

        Private Sub New(kilograms As Double)
            _kilograms = kilograms
        End Sub

        Public ReadOnly Property Kilograms() As Double
            Get
                Return _kilograms
            End Get
        End Property

        Public Shared Function FromKilograms(kilograms As Double) As Mass
            Return New Mass(kilograms)
        End Function

        Public ReadOnly Property SolarMasses() As Double
            Get
                Return _kilograms / KilogramsInSolarMass
            End Get
        End Property

        Public Shared Function FromSolarMasses(solarMasses As Double) As Mass
            Return New Mass(solarMasses * KilogramsInSolarMass)
        End Function

        Public ReadOnly Property EarthMasses() As Double
            Get
                Return _kilograms / KilogramsInEarthMass
            End Get
        End Property

        Public Shared Function FromEarthMasses(earthMasses As Double) As Mass
            Return New Mass(earthMasses * KilogramsInEarthMass)
        End Function

        Public Shadows Function Equals(other As Mass) As Boolean Implements IEquatable(Of Mass).Equals
            If other IsNot Nothing Then
                Return Kilograms.Equals(other.Kilograms)
            End If
            Return False
        End Function
    End Class
End Namespace

File: space-game/Core/Classes/Mass.cs
using System;

namespace Core.Classes
{
    public class Mass : IEquatable<Mass>
    {
        private static readonly double KilogramsInSolarMass = 1.98855 * Math.Pow(10, 30);

        private static readonly double KilogramsInEarthMass = 5973.6 * Math.Pow(10, 21);
        private Mass(double kilograms)
        {
            Kilograms = kilograms;
        }

        public double Kilograms { get; }

        public static Mass FromKilograms(double kilograms)
        {
            return new Mass(kilograms);
        }

        public double SolarMasses => Kilograms / KilogramsInSolarMass;

        public static Mass FromSolarMasses(double solarMasses)
        {
            return new Mass(solarMasses * KilogramsInSolarMass);
        }

        public double EarthMasses => Kilograms / KilogramsInEarthMass;

        public static Mass FromEarthMasses(double earthMasses)
        {
            return new Mass(earthMasses * KilogramsInEarthMass);
        }

        public virtual bool Equals(Mass other)
        {
            return other != null && Kilograms.Equals(other.Kilograms);
        }
    }
}
